### PR TITLE
fix(costmap_generator): restrict costmap within parking lot

### LINKED ISSUE #947: Unreasonable freespace planning area in parking scenario
### Checklist

- [X] I've read the [contribution guidelines](https://github.com/autowarefoundation/autoware/blob/main/CONTRIBUTING.md).
- [X] I've searched other issues and no duplicate issues were found.
- [X] I'm convinced that this is not my fault but a bug.

### Description

In parking scenario as below, the trajectory is extended to lanelets out of the parking lot.
image

### Expected behavior

Trajectory is limited in the parking lot when planning a trajectory to a parking spot using the freespace planner.

### Actual behavior

Trajectory is extended to lanelets out of the parking lot when planning a trajectory to a parking spot using the freespace planner.

### Steps to reproduce

1. Start autoware planning simulator using default
2. Set pose estimate in lanelet within the parking lot
3. Put static obstacles to block the direct way to the target parking spot and then set the goal pose

### Versions

- OS: Ubuntu 20.04
- ROS2: Galactic
- Autoware: Autoware.universe:main

### Possible causes

[CostmapGenerator::loadRoadAreasFromLaneletMap function](https://github.com/autowarefoundation/autoware.universe/blob/1696c2b081f81b43ff2900404e8478dec2cbb1c2/planning/costmap_generator/nodes/costmap_generator/costmap_generator_node.cpp#L231) - target should be limited to the parking area

### Additional context

https://github.com/orgs/autowarefoundation/discussions/317

### PR DESCRIPTION
**Pull request by NorahXiong**
Signed-off-by: NorahXiong <norah.xiong@autocore.ai>

## Description

fix #947 

## Related links

#947 

## Tests performed

Tested in planning simulator.

Before fixing: nearby roads are included in freespace planning boundary
image

After fixing: only parking lot is able for freespace planning
image

## Notes for reviewers

This fixing is only for parking scenario. Not sure if costmap_generator will be used for other scenarios in the future. If so, this fixing may be inappropriate and more comprehensive design is needed.

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [x] The PR follows the [pull request guidelines].
- [x] The PR has been properly tested.
- [x] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.
- [ ] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

File: planning/costmap_generator/nodes/costmap_generator/costmap_generator_node.cpp [hunk 1]
```diff
@@ -276,7 +276,6 @@ void CostmapGenerator::onLaneletMapBin(
   lanelet::utils::conversion::fromBinMsg(*msg, lanelet_map_);
 
   if (use_wayarea_) {
-    loadRoadAreasFromLaneletMap(lanelet_map_, &area_points_);
```
Comment (kenji-miyake) [inline planning/costmap_generator/nodes/costmap_generator/costmap_generator_node.cpp:279]: I think, if you remove this we can't support the case when lanes are not in free space areas like this. (Please regard the red areas are not free space areas.)

Therefore, the vehicle can't plan a trajectory after reaching the blue circle point.
image

Comment (NorahXiong) [reply planning/costmap_generator/nodes/costmap_generator/costmap_generator_node.cpp:279]: I did take this case into consideration but I think the lanes and the parking lot are meant to be overlapped as the comment  of "getLinkedParkingLot" indicates. 
https://github.com/autowarefoundation/autoware.universe/blob/main/map/lanelet2_extension/lib/query.cpp#L437
Not sure about that, if it is not the convention, maybe some guidelines for map producing should be provided.

Besides, it seems no departure strategy is implemented for parking scenario so far. So I think the car can not reach the blue point from a parking spot.

Comment (kenji-miyake) [reply planning/costmap_generator/nodes/costmap_generator/costmap_generator_node.cpp:279]: @NorahXiong Thank you for your comment!

> I did take this case into consideration but I think the lanes and the parking lot are meant to be overlapped as the comment of "getLinkedParkingLot" indicates.

I see. Let me ask other engineers about the map spec. :bow: 
@mitsudome-r @xmfcx @yukkysaito Is it reasonable to assume that lanes are overlapped with free spaces in a parking lot?

> maybe some guidelines for map producing should be provided.

Yes, it is really desired. But there is no good one, we're sorry about that. :cry: 

> Besides, it seems no departure strategy is implemented for parking scenario so far. So I think the car can not reach the blue point from a parking spot.

In this case, I think the vehicle should start with the lane driving scenario.
(If not, I guess it's a bug.)

**Comment by kenji-miyake** [comment]
I'll ask some people who are familiar with the Parking components.

**Comment by kenji-miyake** [comment]
@NorahXiong By the way, could you confirm the guidelines and check these boxes, please?
image

**Comment by NorahXiong** [comment]
> @NorahXiong By the way, could you confirm the guidelines and check these boxes, please? image

Confirmed now.

**Comment by TakaHoribe** [comment]
@NorahXiong 
Thank you for your contribution. 
This module is currently used only for parking purpose but could be extended to any freespace planning usecases. 
So we should have options for users to select the planning area. 
Would you add parameters to select the planning areas in both yaml file and cpp code? It's going to be like:

```yaml
use_wayarea: true
use_parkinglot: true
```

```cpp
if (use_wayarea_) {
    loadRoadAreasFromLaneletMap(lanelet_map_, &area_points_);
}
if (use_parkinglot_) {
    loadParkingAreasFromLaneletMap(lanelet_map_, &area_points_);
}
```

Please let both `true` for now since the map used in the tutorial doesn't have an overlapping area.

**Comment by kenji-miyake** [comment]
> Please let both true for now since the map used in the tutorial doesn't have an overlapping area.

@TakaHoribe Please let me confirm this sentence.
Why did you think "the map used in the tutorial doesn't have an overlapping area"?
I think it does have an overlapping area, and enabling `wayarea` causes this issue.
So I believe `use_wayarea_` should be false. Am I wrong? :thinking:

**Comment by mitsudome-r** [comment]
From the ASWG discussion, @xmfcx and @mitsudome-r agreed to add a parameter as mentioned by @TakaHoribe 's [comment](https://github.com/autowarefoundation/autoware.universe/pull/996#issuecomment-1152289878). 

@NorahXiong Could you modify the code and add a parameter for users to select which map primitives will be used for free space?

**Comment by TakaHoribe** [comment]
@xmfcx Sorry for my late. I'll take it.

**Review by TakaHoribe — APPROVED** [review]
LGTM.

Confirmed works well in my environment.

`use_wayarea = true`
image


`use_wayarea = false`
image

**Comment by TakaHoribe** [comment]
@NorahXiong Sorry for the long wait. Thank you for your work!

**Pull request merged**